compiler/aarch64/codegen/OMRCodeGenerator.cpp

```diff
@@ -19,52 +19,313 @@
+      }
+
+   if (self()->getDebug())
+      self()->getDebug()->startTracingRegisterAssignment();
+
+   while (instructionCursor)
+      {
+      TR::Instruction *prevInstruction = instructionCursor->getPrev();
+
+      self()->tracePreRAInstruction(instructionCursor);
+
+      self()->setCurrentBlockIndex(instructionCursor->getBlockIndex()); // ppc
+
+      instructionCursor->assignRegisters(TR_GPR);
+
+      // Maintain Internal Control Flow Depth
+      // Track internal control flow on labels
+      if (instructionCursor->isLabel())
+         {
+         TR::ARM64LabelInstruction *li = (TR::ARM64LabelInstruction *)instructionCursor;
+
+         if (li->getLabelSymbol() != NULL)
+            {
+            if (li->getLabelSymbol()->isStartInternalControlFlow())
+               {
+               self()->decInternalControlFlowNestingDepth();
+               }
+            if (li->getLabelSymbol()->isEndInternalControlFlow())
+               {
+               self()->incInternalControlFlowNestingDepth();
+               }
+            }
+         }
+
+      self()->freeUnlatchedRegisters();
+      self()->buildGCMapsForInstructionAndSnippet(instructionCursor);
+
+      self()->tracePostRAInstruction(instructionCursor);
+
+      instructionCursor = prevInstruction;
+      }
+
+   if (self()->getDebug())
+      {
+      self()->getDebug()->stopTracingRegisterAssignment();
+      }
+   }
+
+static int32_t identifyFarConditionalBranches(int32_t estimate, TR::CodeGenerator *cg)
+   {
+   TR_Array<TR::ARM64ConditionalBranchInstruction *> candidateBranches(cg->trMemory(), 256);
+   TR::Instruction *cursorInstruction = cg->getFirstInstruction();
+
+   while (cursorInstruction)
+      {
+      TR::ARM64ConditionalBranchInstruction *branch = cursorInstruction->getARM64ConditionalBranchInstruction();
+      if (branch != NULL)
+         {
+         if (abs(branch->getEstimatedBinaryLocation() - branch->getLabelSymbol()->getEstimatedCodeLocation()) > 512 * 1024)
+            {
+            candidateBranches.add(branch);
+            }
+         }
+      cursorInstruction = cursorInstruction->getNext();
+      }
+
+   // The following heuristic algorithm penalizes backward branches in
+   // estimation, since it should be rare that a backward branch needs
+   // a far relocation.
+
+   for (int32_t i = candidateBranches.lastIndex(); i >= 0; i--)
+      {
+      int32_t myLocation = candidateBranches[i]->getEstimatedBinaryLocation();
+      int32_t targetLocation = candidateBranches[i]->getLabelSymbol()->getEstimatedCodeLocation();
+      int32_t j;
+
+      // aarch64 B.cond instruction can reach +/-1MB address range
+
+      if (targetLocation >= myLocation)
+         { // forward branch
+         for (j = i + 1;
+              j < candidateBranches.size() && targetLocation > candidateBranches[j]->getEstimatedBinaryLocation();
+              j++)
+            ;
+
+         if ((targetLocation-myLocation + (j-i-1)*4) >= 1024 * 1024)
```

Comment:
Same comment as above for the materialization of the 1024 constant.

Reply:
There is a comment "aarch64 B.cond instruction can reach +/-1MB address range" in the loop.
Should I define a constant for this to make it clearer?

Reply:
Defined a symbol `AARCH64_BRANCH_REACHABLE_RANGE` for this.